[JENKINS-48761] - Restore binary compatibility with agents running old Remoting versions

oleg-nenashev opened the pull request:
See [JENKINS-48761](https://issues.jenkins-ci.org/browse/JENKINS-48761).

This change reverts new API usages in #3145 and some other places. I made a mistake with this APIs, and my action item is to create a unit test which verifies that the master is at least able to establish connection with an agent with old Remoting version.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Major bug: Restore binary compatibility of Jenkins core logic with old Remoting versions to allow connecting old agents (regression in 2.98)
* ...

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jglick 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

oleg-nenashev committed:
[JENKINS-48761] - Restore binary compatibility with agents running old Remoting versions

jglick commented on core/src/main/java/hudson/Launcher.java:
```diff
@@ -1308,7 +1308,11 @@ public int join() throws InterruptedException, IOException {
                         try {
                             // Sync IO will fail automatically if the channel is being closed, no need to use getOpenChannelOrFail()
-                            taskChannel = Channel.currentOrFail();
+                            // TODOL Replace by Channel#currentOrFail() when Remoting version allows
```
typo

oleg-nenashev commented on core/src/main/java/jenkins/security/MasterToSlaveCallable.java:
```diff
@@ -11,10 +17,31 @@
+    protected static Channel _getOpenChannelOrFail() throws ChannelClosedException {
+        final Channel ch = _getChannelOrFail();
+        if (ch.isClosingOrClosed()) { // TODO: Since Remoting 2.33, we still need to explicitly declare minimal Remoting version
+            throw new ChannelClosedException("The associated channel " + ch + " is closing down or has closed down", ch.getCloseRequestCause());
```
This one is 3.11, will need an older API as well

jglick commented on core/src/main/java/jenkins/security/MasterToSlaveCallable.java:
```diff
@@ -11,10 +17,31 @@
 
-    private static final long serialVersionUID = 1L;
+    //TODO: replace by Callable#getChannelOrFail() once Minimal supported Remoting version is 3.15 or above
+    @Restricted(NoExternalUse.class)
```
I think you could do something like

```java
@Override
public Channel getChannelOrFail() throws ChannelStateException {
    // copy implementation from Callable
}
```

which would avoid polluting the API with even a restricted function, while still fixing the problem.

oleg-nenashev replied:
I was not sure it's 100% safe for defaults and decided not to try due to the urgency. Can try it now since I have a test which should reveal the problem if any

jglick replied:
Not at all urgent, just an idea which might simplify the patch.

oleg-nenashev committed:
[JENKINS-48761] - Use pre-Remoting 3.0 ChannelClosedException APIs

oleg-nenashev commented:
The current code will run on Remoting agents 3.0+ at least, likely on even older ones

jglick reviewed (dismissed):
OK as a hotfix.

The broader issue is that we have no static check for the case that a core or plugin remote `Callable` is using an API member introduced in a newer version of Remoting than the agent uses. Not sure what to do about that.

oleg-nenashev commented:
I could write a FindBugs detector for that, but it's hard do guarantee full
coverage. Maybe we should start from playing with createOnlineSlave() for
old Remoting versions and duplicate some tests

On Jan 2, 2018 22:28, "Jesse Glick" <notifications@github.com> wrote:

> *@jglick* approved this pull request.
>
> OK as a hotfix.
>
> The broader issue is that we have no static check for the case that a core
> or plugin remote Callable is using an API member introduced in a newer
> version of Remoting than the agent uses. Not sure what to do about that.
> ------------------------------
>
> In core/src/main/java/hudson/Launcher.java
> <https://github.com/jenkinsci/jenkins/pull/3212#discussion_r159317411>:
>
> > @@ -1308,7 +1308,11 @@ public int join() throws InterruptedException, IOException {
>                          Channel taskChannel = null;
>                          try {
>                              // Sync IO will fail automatically if the channel is being closed, no need to use getOpenChannelOrFail()
> -                            taskChannel = Channel.currentOrFail();
> +                            // TODOL Replace by Channel#currentOrFail() when Remoting version allows
>
> typo
> ------------------------------
>
> In core/src/main/java/jenkins/security/MasterToSlaveCallable.java
> <https://github.com/jenkinsci/jenkins/pull/3212#discussion_r159318752>:
>
> >      @Override
>      public void checkRoles(RoleChecker checker) throws SecurityException {
>          checker.check(this,Roles.SLAVE);
>      }
>
> -    private static final long serialVersionUID = 1L;
> +    //TODO: replace by Callable#getChannelOrFail() once Minimal supported Remoting version is 3.15 or above
> +    @Restricted(NoExternalUse.class)
>
> I think you could do something like
>
> @Overridepublic Channel getChannelOrFail() throws ChannelStateException {
>     // copy implementation from Callable
> }
>
> which would avoid polluting the API with even a restricted function, while
> still fixing the problem.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/3212#pullrequestreview-86246556>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC3IoGnbQ969x0_Pq77u_iEObMqsA7Ekks5tGp-DgaJpZM4RRFc3>
> .
>

oleg-nenashev commented:
@jenkinsci/code-reviewers does it deserve an out-of-order release? I'd guess so, but maybe it makes sense to take another day for test automation, etc.

jglick commented:
> playing with createOnlineSlave() for old Remoting versions

Would be tricky since you `remoting.jar` is picked up from the Maven test classpath, which cannot contain multiple versions. Better would be to use `docker-fixtures` from `jenkinsci/jenkins/test/`, which would allow us to connect to an actual agent running in another environment and using a pinned version of Remoting. (For that matter, did none of the existing tests using Dockerized SSH agents in `acceptance-test-harness` catch this problem?)

> does it deserve an out-of-order release?

I am not sure. It only affects people running old agent versions. If they are going to bother updating core to get an out-of-order release, maybe they could just as easily update their agent. No strong opinion.

oleg-nenashev commented:
> For that matter, did none of the existing tests using Dockerized SSH agents in acceptance-test-harness catch this problem?

Let's see. But we do not run ATH before weeklies now

oleg-nenashev committed:
[JENKINS-48766] - Create restricted API for getting info about Remoting versions

oleg-nenashev committed:
[JENKINS-48761] - Create a smoke test for old agents

oleg-nenashev commented:
@jglick After some consideration I decided to write a test without Docker containers. It required doing some foundation work for [JENKINS-48766](https://issues.jenkins-ci.org/browse/JENKINS-48766), but I think we need to do it anyway.

oleg-nenashev commented on test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    @Test
+    @Issue("JENKINS-48761")
```
This test fails on Jenkins 2.99/98. `ChannelClosedException` API compatibility is not tested though

oleg-nenashev commented on pom.xml:
```diff
@@ -104,6 +104,10 @@ THE SOFTWARE.
 
     <maven-war-plugin.version>3.0.0</maven-war-plugin.version> <!-- JENKINS-47127 bump when 3.2.0 is out. Cf. MWAR-407 -->
 
+    <!-- Minimal Remoting version, which is tested for API compatibility -->
```
Just picked first version with excessive stability improvements and available changelog: https://github.com/jenkinsci/remoting/blob/master/CHANGELOG.md#260 . This version has been released on June 10, 2016 => old enough.

IMHO the version should be bumped to 3.0 once https://issues.jenkins-ci.org/browse/JENKINS-48766 is implemented. It would unblock #3188 from @Vlatombe

oleg-nenashev commented:
@reviewbybees Needs re-reviews (mostly tests and Restricted logic in the core for enabling such kind of tests)

jglick commented on core/src/filter/resources/jenkins/slaves/remoting-info.properties:
```diff
@@ -0,0 +1,6 @@
+# Remoting version, which is embedded into the core
+# This version MAY differ from what is really classloaded (see the "Pluggable Core Components", JENKINS-41196)
+remoting.embedded.version=${remoting.version}
```
Or this could be computed from `remoting*.jar!/META-INF/MANIFEST.MF#Version` using `Which`.

oleg-nenashev replied:
I would prefer to keep it as is. 
 
* At some point we may want to write a "Remoting Compatibility Tester" tool or so
* Not sure how which is going to behave once we have pluggable core components in place

jglick commented on core/src/filter/resources/jenkins/slaves/remoting-info.properties:
```diff
@@ -0,0 +1,6 @@
+remoting.embedded.version=${remoting.version}
+
+# Minimal Remoting version on external agents which is supported by the core
+remoting.minimal.supported.version=${remoting.minimal.supported.version}
```
s/minimal/minimum/g

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+
+// TODO: Make the API public (JENKINS-48766)
+/**
+ * Provides information about Remoting versions used withing the core.
```
typo

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Provides information about Remoting versions used withing the core.
+ * @author Oleg Nenashev
+ * @since TODO
```
Restricted, so delete.

oleg-nenashev replied:
There is a TODO for making it public above

jglick replied:
So if and when that happens, you will add a `@since` with _that_ version, as with adding any new API. We should only have `@since TODO` comments in the code base for those things which we have forgotten to fill in after merging.

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+    static {
+        Properties props = new Properties();
+        try (InputStream is = RemotingVersionInfo.class.getResourceAsStream(RESOURCE_NAME)) {
+            if(is!=null) {
```
No need to check for null. If the file is missing, core is corrupt, so die now.

oleg-nenashev replied:
The implementation is similar to `jenkins.model.Jenkins`'s appoarch to reading `jenkins-version.properties`. I agree this should not be a problem, but there may be edge cases (e.g. custom packaging). I see no problem with having a fault-tolerant code since it is a regression fix

jglick replied:
In the case of custom packaging, it is the responsibility of the packager to not make mistakes. If they do, their WAR will be broken. We should not engineer in robustness for something that is not a user error and is trivially detected.

oleg-nenashev replied:
Hmm, another potential use-case is calling this class from the agent. I doubt we would be dare enough to send Jenkins core over Remoting. Do we do that in the case of such requests? I hope not

oleg-nenashev replied:
Let's keep it as is. I will improve checks when exposing it to public API

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+                props.load(is);
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.WARNING, "Failed to load Remoting Info from " + RESOURCE_NAME, e);
```
Ditto—just rethrow as `ExceptionInInitializerError`.

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+    private static VersionNumber tryExtractVersion(@Nonnull Properties props, @Nonnull String propertyName) {
+        String prop = props.getProperty(propertyName);
+        if (prop == null) {
+            LOGGER.log(Level.FINE, "Property {0} is not defined in {1}", new Object[] {propertyName, RESOURCE_NAME});
```
Throw an `IllegalStateException`.

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        if(prop.contains("${")) { // Due to whatever reason, Maven does not nullify them
+            LOGGER.log(Level.WARNING, "Property {0} in {1} has unresolved variable(s). Raw value: {2}",
```
ditto

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+        try {
+            return new VersionNumber(prop);
+        } catch (RuntimeException ex) {
+            LOGGER.log(Level.WARNING, String.format("Failed to parse version for for property %s in %s. Raw Value: %s",
```
Ditto. We are in full control of this file, so there is no reason to tolerate any errors.

jglick commented on core/src/main/java/jenkins/slaves/RemotingVersionInfo.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    @CheckForNull
```
`@Nonnull` since we know these will be defined.

jglick commented on core/src/test/java/jenkins/slaves/RemotingVersionInfoTest.java:
```diff
@@ -0,0 +1,47 @@
+    @Test
+    public void shouldLoadEmbeddedVersionByDefault() {
+        assertThat("Remoting Embedded version is not defined",
+                RemotingVersionInfo.getEmbeddedVersion(), notNullValue());
```
Unnecessary if they are just `@Nonnull` to begin with, since other tests would die if someone screwed this up.

jglick commented on test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java:
```diff
@@ -0,0 +1,140 @@
+    @Before
+    public void extractAgent() throws Exception {
+        agentJar = new File(tmpDir.getRoot(), "old-agent.jar");
+        FileUtils.copyURLToFile(getClass().getResource("/old-remoting/remoting-minimal-supported.jar"), agentJar);
```
BTW as a stylistic matter it is safest to use `OldRemotingAgentTest.class` in place of `getClass()`. (Unless the test suite is `final` in which case they are equivalent.)

oleg-nenashev replied:
yup

jglick commented on test/src/test/java/jenkins/slaves/OldRemotingAgentTest.java:
```diff
@@ -0,0 +1,140 @@
+            try {
+                monitorMethod = AbstractAsyncNodeMonitorDescriptor.class.getDeclaredMethod("monitor", Computer.class);
+            } catch (NoSuchMethodException ex) {
+                System.out.println("Cannot invoke monitor " + monitor + ", no monitor(Computer.class) method in the Descriptor. It will be ignored. " + ex.getMessage());
```
This is test code. Just throw it up!

Or, better, make that method `@Restricted(NoExternalUse.class) public`.

oleg-nenashev replied:
Right, it should never happen

oleg-nenashev committed:
[JENKINS-48761] - Address review comments from @jglick

oleg-nenashev commented:
@jglick addressed some comments. One biggest open question is `RemotingVersionInfo`. You propose to make the checks stricter and to throw exceptions on initialization errors. I kinda agree with that, but I propose to postpone it and to handle it in https://issues.jenkins-ci.org/browse/JENKINS-48766 while making the API public. 

Now it's used only for tests, and I would prefer to keep it permissive in order to avoid new regressions.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -32,7 +32,6 @@
 import hudson.slaves.Cloud;
 import jenkins.util.SystemProperties;
 import hudson.Util;
-import hudson.cli.declarative.CLIMethod;
```
unrelated but OK

jglick commented on test/pom.xml:
```diff
@@ -224,6 +224,31 @@ THE SOFTWARE.
+                  <version>${remoting.minimum.supported.version}</version>
+                  <type>jar</type>
+                  <outputDirectory>${project.build.outputDirectory}/old-remoting</outputDirectory>
+                  <destFileName>remoting-minimal-supported.jar</destFileName>
```
BTW may want to s/minimal/minimum/ here too for consistency.

oleg-nenashev replied:
Will address it in a follow-up

oleg-nenashev commented:
@reviewbybees done
@jenkinsci/code-reviewers I kindly ask for approval to got forward with merge and out-of-order release. Community stats are bad due to that regression:

![screen shot 2018-01-03 at 19 24 19](https://user-images.githubusercontent.com/3000480/34541586-19b5037a-f0d9-11e7-9889-f40bee703c8c.png)

oleg-nenashev commented:
Merging it, this change will unlikely make the situation worse

oleg-nenashev commented:
Retrospective thread: https://groups.google.com/forum/#!topic/jenkinsci-dev/-VQ5YU0yYhg
